File: nahomy-c2-backend/src/main/container.py
from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker
from ..infrastructure.persistence.database import engine, async_session_factory
from ..infrastructure.persistence.models import Base
from ..infrastructure.persistence.repositories import (
    SQLAlchemyUsuarioRepository,
    SQLAlchemyProcesoIntencionRepository,
    SQLAlchemyEjecucionAutomatizadaRepository,
    SQLAlchemyFeedbackReporteRepository,
    SQLAlchemyLogAuditoriaRepository,
    SQLAlchemyMetricaRendimientoRepository,
)
from ..infrastructure.messaging.rabbitmq_adapter import RabbitMQAdapter
from ..infrastructure.webserver.fastapi_server import create_app
from ..infrastructure.config.rabbitmq_config import get_rabbitmq_url
from ..application.use_cases import (
    AutenticarUsuarioUseCase,
    IniciarIntencionUseCase,
    EjecutarComandoUseCase,
    GenerarFeedbackUseCase,
    ObtenerMetricasUseCase,
    ObtenerLogsAuditoriaUseCase,
)
from ..interface_adapters.message_handlers import IntencionMessageHandler


class Container:
    def __init__(self):
        self._session_factory = async_session_factory
        self._message_bus: RabbitMQAdapter | None = None
        self.app = None

    async def initialize(self):
        async with engine.begin() as conn:
            await conn.run_sync(Base.metadata.create_all)

        self._message_bus = RabbitMQAdapter(get_rabbitmq_url())
        await self._message_bus.connect()

        async def get_session() -> AsyncSession:
            async with self._session_factory() as session:
                yield session

        repos = {}
        async with self._session_factory() as session:
            repos["usuario"] = SQLAlchemyUsuarioRepository(session)
            repos["proceso"] = SQLAlchemyProcesoIntencionRepository(session)
            repos["ejecucion"] = SQLAlchemyEjecucionAutomatizadaRepository(session)
            repos["feedback"] = SQLAlchemyFeedbackReporteRepository(session)
            repos["log"] = SQLAlchemyLogAuditoriaRepository(session)
            repos["metrica"] = SQLAlchemyMetricaRendimientoRepository(session)

        auth_uc = AutenticarUsuarioUseCase(repos["usuario"])
        intencion_uc = IniciarIntencionUseCase(
            repos["proceso"], repos["usuario"], self._message_bus
        )
        comando_uc = EjecutarComandoUseCase(
            repos["proceso"], repos["ejecucion"], self._message_bus
        )
        feedback_uc = GenerarFeedbackUseCase(
            repos["feedback"], repos["metrica"], self._message_bus
        )
        metricas_uc = ObtenerMetricasUseCase(repos["metrica"])
        logs_uc = ObtenerLogsAuditoriaUseCase(repos["log"])

        handler = IntencionMessageHandler(self._message_bus)
        self._message_bus.subscribe(
            "c2.intention.*", handler.handle_intencion_iniciada
        )
        self._message_bus.subscribe(
            "c2.command.*", handler.handle_comando_ejecutado
        )

        self.app = create_app(
            auth_uc, intencion_uc, comando_uc, metricas_uc, logs_uc, feedback_uc
        )

        print("[Container] Inicializado correctamente")

    async def shutdown(self):
        if self._message_bus:
            await self._message_bus.disconnect()
        await engine.dispose()
        print("[Container] Recursos liberados")

File: nahomy-c2-backend/src/main/container.ts
import { DataSource } from 'typeorm'
import { createDataSource } from '../infrastructure/config/DatabaseConfig'
import { RabbitMQConfig } from '../infrastructure/config/RabbitMQConfig'
import { RabbitMQAdapter } from '../infrastructure/messaging/RabbitMQAdapter'
import { ExpressServer } from '../infrastructure/webserver/ExpressServer'

import { TypeORMUsuarioRepository } from '../infrastructure/persistence/typeorm/repositories/TypeORMUsuarioRepository'
import { TypeORMProcesoIntencionRepository } from '../infrastructure/persistence/typeorm/repositories/TypeORMProcesoIntencionRepository'
import { TypeORMEjecucionAutomatizadaRepository } from '../infrastructure/persistence/typeorm/repositories/TypeORMEjecucionAutomatizadaRepository'
import { TypeORMFeedbackReporteRepository } from '../infrastructure/persistence/typeorm/repositories/TypeORMFeedbackReporteRepository'
import { TypeORMLogAuditoriaRepository } from '../infrastructure/persistence/typeorm/repositories/TypeORMLogAuditoriaRepository'
import { TypeORMMetricaRendimientoRepository } from '../infrastructure/persistence/typeorm/repositories/TypeORMMetricaRendimientoRepository'

import { AutenticarUsuarioUseCase } from '../application/use-cases/AutenticarUsuarioUseCase'
import { IniciarIntencionUseCase } from '../application/use-cases/IniciarIntencionUseCase'
import { EjecutarComandoUseCase } from '../application/use-cases/EjecutarComandoUseCase'
import { GenerarFeedbackUseCase } from '../application/use-cases/GenerarFeedbackUseCase'
import { ObtenerMetricasUseCase } from '../application/use-cases/ObtenerMetricasUseCase'
import { ObtenerLogsAuditoriaUseCase } from '../application/use-cases/ObtenerLogsAuditoriaUseCase'

import { AuthController } from '../interface-adapters/controllers/AuthController'
import { IntencionController } from '../interface-adapters/controllers/IntencionController'
import { ComandoController } from '../interface-adapters/controllers/ComandoController'
import { MetricasController } from '../interface-adapters/controllers/MetricasController'
import { FeedbackController } from '../interface-adapters/controllers/FeedbackController'
import { IntencionMessageHandler } from '../interface-adapters/message-handlers/IntencionMessageHandler'

export class Container {
  private _dataSource!: DataSource
  private _messageBus!: RabbitMQAdapter
  private _server!: ExpressServer

  async initialize(): Promise<void> {
    this._dataSource = createDataSource()
    await this._dataSource.initialize()
    console.log('[DB] Conexión a PostgreSQL establecida')

    const rabbitConfig = new RabbitMQConfig()
    this._messageBus = new RabbitMQAdapter(rabbitConfig)
    await this._messageBus.connect()

    const usuarioRepo = new TypeORMUsuarioRepository(this._dataSource.getRepository('UsuarioEntity'))
    const procesoRepo = new TypeORMProcesoIntencionRepository(this._dataSource.getRepository('ProcesoIntencionEntity'))
    const ejecucionRepo = new TypeORMEjecucionAutomatizadaRepository(this._dataSource.getRepository('EjecucionAutomatizadaEntity'))
    const feedbackRepo = new TypeORMFeedbackReporteRepository(this._dataSource.getRepository('FeedbackReporteEntity'))
    const logRepo = new TypeORMLogAuditoriaRepository(this._dataSource.getRepository('LogAuditoriaEntity'))
    const metricaRepo = new TypeORMMetricaRendimientoRepository(this._dataSource.getRepository('MetricaRendimientoEntity'))

    const authUseCase = new AutenticarUsuarioUseCase(usuarioRepo)
    const iniciarIntencionUseCase = new IniciarIntencionUseCase(procesoRepo, usuarioRepo, this._messageBus)
    const ejecutarComandoUseCase = new EjecutarComandoUseCase(procesoRepo, ejecucionRepo, this._messageBus)
    const generarFeedbackUseCase = new GenerarFeedbackUseCase(feedbackRepo, metricaRepo, this._messageBus)
    const obtenerMetricasUseCase = new ObtenerMetricasUseCase(metricaRepo)
    const obtenerLogsUseCase = new ObtenerLogsAuditoriaUseCase(logRepo)

    const authController = new AuthController(authUseCase)
    const intencionController = new IntencionController(iniciarIntencionUseCase)
    const comandoController = new ComandoController(ejecutarComandoUseCase)
    const metricasController = new MetricasController(obtenerMetricasUseCase, obtenerLogsUseCase)
    const feedbackController = new FeedbackController(generarFeedbackUseCase)

    const messageHandler = new IntencionMessageHandler(this._messageBus)
    this._messageBus.subscribe('c2.intention.*', (event) => messageHandler.handleIntencionIniciada(event))
    this._messageBus.subscribe('c2.command.*', (event) => messageHandler.handleComandoEjecutado(event))

    this._server = new ExpressServer(
      authController,
      intencionController,
      comandoController,
      metricasController,
      feedbackController,
    )
    this._server.setup()

    const port = Number(process.env.PORT) || 3000
    this._server.start(port)
  }

  async shutdown(): Promise<void> {
    await this._messageBus?.disconnect()
    await this._dataSource?.destroy()
    console.log('[Container] Recursos liberados')
  }
}
